net/libfab/libfab.c

```diff
@@ -0,0 +1,3455 @@
+ *       the buffer times out (ma_buf_timeout())) and synchronously (when the
+ *       buffer is canceled by the user (m0_net_buf_del()->...->buf_del()). The
+ *       difficulty is that net.h interface specifies that the transfer machine
+ *       lock is to be released before invoking the call-back. This cannot be
+ *       done in a synchronous context (to avoid breaking invariants), so in
+ *       this case the buffer is queued to a special ma::t_done queue which is
+ *       processed asynchronously by ma_buf_done().
+ *
+ * Libfabric interface:
+ * --------------------
+ *
+ * Libfabric module uses fi_endpoint() to create transport level communication
+ * portals. Libfabric supports two types of endpoints: Passive endpoint and
+ * Active endpoint. Passive endpoints are most often used to listen for
+ * incoming connection requests. However, a passive endpoint may be used to
+ * reserve a fabric address that can be granted to an active endpoint.
+ * Active endpoints belong to access domains and can perform data transfers.
+ *
+ * Below Libfabric interfaces are used:
+ * 1) fi_fabric : Used for fabric domain operations.
+ *    Reference: https://ofiwg.github.io/libfabric/v1.6.2/man/fi_fabric.3.html
+ *
+ * 2) fi_domain: Used for accessing fabric domain.
+ *    Reference: https://ofiwg.github.io/libfabric/v1.1.1/man/fi_domain.3.html
+ *
+ * 3) fi_endpoint: USed for fabric endpoint operations.
+ *    Reference: https://ofiwg.github.io/libfabric/master/man/fi_endpoint.3.html
+ *
+ * 4) fi_mr: Used for memory region operations.
+ *    Reference: https://ofiwg.github.io/libfabric/v1.1.1/man/fi_mr.3.html
+ *
+ * 5) fi_rma: Used for remote memory access operations.
+ *    Reference: https://ofiwg.github.io/libfabric/v1.1.1/man/fi_rma.3.html
+ *
+ * @{
+ */
+
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_NET
+#include "lib/trace.h"          /* M0_ENTRY() */
+
+#ifdef ENABLE_LIBFAB
+
+#include <netinet/in.h>         /* INET_ADDRSTRLEN */
+#include <arpa/inet.h>          /* inet_pton, htons */
+#include <sched.h>              /* sched_yield */
+#include <stdlib.h>             /* atoi */
+#include <sys/epoll.h>          /* struct epoll_event */
+#include <unistd.h>             /* close */
+#include "net/buffer_pool.h"    /* struct m0_net_buffer_pool */
+#include "net/net.h"            /* struct m0_net_domain */
+#include "lib/errno.h"          /* errno */
+#include "lib/finject.h"        /* M0_FI_ENABLED */
+#include "lib/hash.h"           /* m0_htable */
+#include "lib/memory.h"         /* M0_ALLOC_PTR()*/
+#include "libfab_internal.h"
+#include "lib/string.h"         /* m0_streq */
+#include "net/net_internal.h"   /* m0_net__buffer_invariant() */
+
+#define U32_TO_VPTR(a)     ((void*)((uintptr_t)a))
+#define VPTR_TO_U32(a)     ((uint32_t)((uintptr_t)a))
+
+static char     *providers[] = { "verbs", "tcp", "sockets" };
+static char     *protf[]  = { "unix", "inet", "inet6" };
```

Comment:
This was only added to support sock addressing format. This will be cleaned up along with address format changes task where we will support a common format for sock and libfabric
<xprt (sock/libfab)>:<family (inet/inet6)>:<type (tcp/verbs)>:<IP/FQDN>@<port>